### ISSUE TITLE
Document why the complexity should be high (or not)

### ISSUE THREAD

**Issue by mnil2** [slarse/lektor#85]
Include here a comment documenting "why the complexity should be high (or not)" for both your assigned functions.

**Comment by larshum** [comment]
* `utils.prune_file_and_folder`:
The complexity should be lower since it could quite easily be divided into several functions.
* `utils.decode_flat_data`:
This function is already divided into a few inner functions. It still has a high CC, and it seems like it would be hard to decrease it further without a major refactoring.

**Comment by rouwaydhanna** [comment]
#### `utils.get_structure_hash`
This functions essentially calls another function `_hash` defined inside it doing all the work. In fact, `get_structure_hash` has a CC of 1 but all it does is call `_hash` and return whatever it returns. If the inside function was included, `utils.get_structure_hash` has a high CC of 14, which is plausible since  `_hash` contains 10 if statements and 3 loops. This means a lot of decision making and branching.

#### `utils.locate_executable`
This function tries to locate the path to an executable. Sounds simple enough but there are some factors involved in deciding where to look for that executable and what extensions to use, for example whether the parameter given is a name of an executable of a path. These come with some decision making and looping, which increases the complexity. Also, the type of os that is running the function affects the behaviour of it which increases the complexity since it adds another decision.

**Comment by mnil2** [comment]
* `lektor.sourcesearch.find_files`:
The complexity is almost all in managing the default values for the various file property filters. For example, if no language is provided, the language defaults to English. I think this particular functionality should be handled somewhere else, and the default filter values be handled by a configuration rather than hardcoded in the function.

* `lektor.types.flow.discover_relevant_flowblock_models`:
This function appears just a little too complex for its purpose for two reasons. One, querying for all the flowblocks instead of a subset results in a completely separate branch from a normal query, resulting in some minimal code duplication. Two, the special case handling of nested flowblocks alone results in an increased CC of 4, but this is a trivially separable function.

* `lektor.db._CmpHelper.coerce`:
This function is probably as complex as it needs to be. It attempts to coerce two values such that they are comparable, which necessarily involves handling a lot of cases. Further, it's a specialist function that doesn't try to do anything else. Maybe there's a cooler way of doing it.

**Comment by mudaliarkartik** [comment]
- `utils.merge`
Complexity is of this function is high because, branching is happening based on datatype of input parameters and merge is recursive function, so based on the input the merge function complexity changes. Merge function is handling lists and dictionary data types and that makes the complexity high because branching is required to handle both data types.

- `metaformat.tokenize`
Tokenize function has branching based on the inputs i.e iterable, encoding and keys. So based on different values of encoding and keys branching occurs which makes the function complex.
Tokenize also makes function calls to `_line_is_dashes` and `_process_buf` while executing which increases the complexity. This high complexity is required so that tokenize function can handle different type of data with filtering

**Comment by slarse** [comment]
* `imagetools.get_image_info`
This function really has no business being this complex (CC=14). It is entirely dominated both in terms of line count and CC by parsing the `JPEG` header info, which can just as easily be put into its own function. This is one of the functions that we chose to refactor, and with little effort we broke it into three functions with CCs if 3, 5 and 8.

* `cli.content_file_info_cmd`
This is yet another function that is doing too many things. It essentially has four disconnected steps that can easily be factored out. There is both a refactor plan, and an actual refactor of it, so there is little value in writing more about it here.
